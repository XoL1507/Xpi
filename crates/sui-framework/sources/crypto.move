// Copyright (c) 2022, Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

/// Library for cryptography onchain.
module sui::crypto {
    friend sui::validator;

    /// @param signature: A 65-bytes signature in form (r, s, v) that is signed using 
    /// Secp256k1. Reference implementation on signature generation using RFC6979: 
    /// https://github.com/MystenLabs/narwhal/blob/5d6f6df8ccee94446ff88786c0dbbc98be7cfc09/crypto/src/secp256k1.rs
    /// The accepted v values are {0, 1, 2, 3}.
    /// 
    /// @param hashed_msg: the hashed 32-bytes message. The message must be hashed instead 
    /// of plain text to be secure.
    /// 
    /// If the signature is valid, return the corresponding recovered Secpk256k1 public 
    /// key, otherwise throw error. This is similar to ecrecover in Ethereum, can only be 
    /// applied to Secp256k1 signatures.
    public native fun ecrecover(signature: vector<u8>, hashed_msg: vector<u8>): vector<u8>;

    /// @param data: arbitrary bytes data to hash
    /// Hash the input bytes using keccak256 and returns 32 bytes.
    public native fun keccak256(data: vector<u8>): vector<u8>;

    /// @param signature: A 48-bytes signature that is a point on the G1 subgroup
    /// @param public_key: A 96-bytes public key that is a point on the G2 subgroup
    /// @param msg: The message that we test the signature against.
    ///
    /// If the signature is a valid BLS12381 signature of the message and public key, return true.
    /// Otherwise, return false.
    public native fun bls12381_verify_g1_sig(signature: vector<u8>, public_key: vector<u8>, msg: vector<u8>): bool; 

    /// @param signature: A 65-bytes signature in form (r, s, v) that is signed using 
    /// Secp256k1. Reference implementation on signature generation using RFC6979: 
    /// https://github.com/MystenLabs/narwhal/blob/5d6f6df8ccee94446ff88786c0dbbc98be7cfc09/crypto/src/secp256k1.rs
    /// 
    /// @param public_key: The public key to verify the signature against
    /// @param hashed_msg: The hashed 32-bytes message, same as what the signature is signed against.
    /// 
    /// If the signature is valid to the pubkey and hashed message, return true. Else false.
    public native fun secp256k1_verify(signature: vector<u8>, public_key: vector<u8>, hashed_msg: vector<u8>): bool;

    use sui::elliptic_curve::{Self as ec, RistrettoPoint};

    /// Only bit_length = 64, 32, 16, 8 will work.
    native fun native_verify_full_range_proof(proof: vector<u8>, commitment: vector<u8>, bit_length: u64);

    /// @param proof: The bulletproof
    /// @param commitment: The commitment which we are trying to verify the range proof for
    /// @param bit_length: The bit length that we prove the committed value is whithin. Note that bit_length must be either 64, 32, 16, or 8.
    /// 
    /// If the range proof is valid, execution succeeds, else panics.
    public fun verify_full_range_proof(proof: vector<u8>, commitment: RistrettoPoint, bit_length: u64) {
        native_verify_full_range_proof(proof, ec::bytes(&commitment), bit_length)
    }

    /// @param signature: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param public_key: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param msg: The message that we test the signature against.
    /// 
    /// If the signature is a valid Ed25519 signature of the message and public key, return true.
    /// Otherwise, return false.
    public(friend) native fun ed25519_verify(signature: vector<u8>, public_key: vector<u8>, msg: vector<u8>): bool;

    /// @param signature: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param public_key: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param msg: The message that we test the signature against.
    /// @param domain: The domain that the signature is tested again. We essentially prepend this to the message.
    /// 
    /// If the signature is a valid Ed25519 signature of the message and public key, return true.
    /// Otherwise, return false.
    public(friend) fun ed25519_verify_with_domain(signature: vector<u8>, public_key: vector<u8>, msg: vector<u8>, domain: vector<u8>): bool {
        std::vector::append(&mut domain, msg);
        ed25519_verify(signature, public_key, domain)
    }
}
