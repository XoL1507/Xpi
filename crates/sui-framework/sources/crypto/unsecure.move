// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

module sui::unsecure {
    friend sui::validator;

    public native fun unsecure_verify(signature: &vector<u8>, public_key: &vector<u8>, msg: &vector<u8>): bool;

    public(friend) fun unsecure_verify_with_domain(
        signature: &vector<u8>,
        public_key: &vector<u8>,
        msg: vector<u8>,
        domain: vector<u8>
    ): bool {
        std::vector::append(&mut domain, msg);
        unsecure_verify(signature, public_key, &domain)
    }

}
