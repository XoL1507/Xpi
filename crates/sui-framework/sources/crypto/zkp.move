// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

module sui::zkp {
    use std::vector;

    /// Length of the vector<u8> representing a SHA3-256 digest.
    const LENGTH: u64 = 32;

    /// Error code when the length is invalid.
    const LengthMismatch: u64 = 0;

    /// Proof<Bls12_381>
    struct Proof has store, copy, drop {
        bytes: vector<u8>
    }

    /// PreparedVerifyingKey
    struct PreparedVerifyingKey has store, copy, drop {
        /// The element vk.gamma_abc_g1,
        /// aka the `[gamma^{-1} * (beta * a_i + alpha * b_i + c_i) * G]`, where i spans the public inputs
        vk_gamma_abc_g1: vector<u8>,
        /// The element `e(alpha * G, beta * H)` in `E::GT`. blst_fp12
        alpha_g1_beta_g2: vector<u8>,
        /// The element `- gamma * H` in `E::G2`, for use in pairings.
        gamma_g2_neg_pc: vector<u8>,
        /// The element `- delta * H` in `E::G2`, for use in pairings.
        delta_g2_neg_pc: vector<u8>,
    }

    public fun pvk_from_bytes(
        vk_gamma_abc_g1: vector<u8>, 
        alpha_g1_beta_g2: vector<u8>, 
        gamma_g2_neg_pc: vector<u8>, 
        delta_g2_neg_pc: vector<u8>): PreparedVerifyingKey {
        PreparedVerifyingKey { vk_gamma_abc_g1, alpha_g1_beta_g2, gamma_g2_neg_pc, delta_g2_neg_pc }
    }
    
    public fun proof_from_bytes(bytes: vector<u8>): Proof {
        Proof { bytes }
    }
    /// @param pvk: PreparedVerifyingKey
    ///
    /// @param x
    ///
    /// @param proof
    /// Returns the validity of the Groth16 proof passed as argument.
    public fun verify_groth16_proof(pvk: PreparedVerifyingKey, x: vector<u8>, proof: Proof): bool {
        internal_verify_groth16_proof(
            pvk.vk_gamma_abc_g1, 
            pvk.alpha_g1_beta_g2, 
            pvk.gamma_g2_neg_pc, 
            pvk.delta_g2_neg_pc, 
            x, 
            proof.bytes
        )
    }

    public native fun internal_verify_groth16_proof(
        vk_gamma_abc_g1_bytes: vector<u8>, 
        alpha_g1_beta_g2_bytes: vector<u8>, 
        gamma_g2_neg_pc_bytes: vector<u8>, 
        delta_g2_neg_pc_bytes: vector<u8>, 
        x_bytes: vector<u8>, 
        proof_bytes: vector<u8>
    ): bool;
}
