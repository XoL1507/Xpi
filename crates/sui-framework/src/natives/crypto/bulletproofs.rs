// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0
use crate::legacy_empty_cost;
use fastcrypto::{
    bulletproofs::{BulletproofsRangeProof, PedersenCommitment},
    traits::ToFromBytes,
};
use move_binary_format::errors::PartialVMResult;
use move_vm_runtime::native_functions::NativeContext;
use move_vm_types::{
    loaded_data::runtime_types::Type,
    natives::function::NativeResult,
    pop_arg,
    values::{Value, VectorRef},
};
use smallvec::smallvec;
use std::collections::VecDeque;

pub const INVALID_BULLETPROOF: u64 = 0;
pub const INVALID_RISTRETTO_GROUP_ELEMENT: u64 = 1;

/// Using the word "sui" for nothing-up-my-sleeve number guarantees.
pub const BP_DOMAIN: &[u8] = b"sui";

pub fn verify_range_proof(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 3);

    let bit_length = pop_arg!(args, u64);
    let commitment_bytes = pop_arg!(args, VectorRef);
    let proof_bytes = pop_arg!(args, VectorRef);
    let cost = legacy_empty_cost();

    let commitment_bytes_ref = commitment_bytes.as_bytes_ref();
    let proof_bytes_ref = proof_bytes.as_bytes_ref();

    let Ok(proof) = BulletproofsRangeProof::from_bytes(&proof_bytes_ref) else { return Ok(NativeResult::err(cost, INVALID_BULLETPROOF)); };

    let Ok(commitment) = PedersenCommitment::from_bytes(&commitment_bytes_ref) else { return Ok(NativeResult::err(cost, INVALID_RISTRETTO_GROUP_ELEMENT)); };

    match proof.verify_bit_length(&commitment, bit_length as usize, BP_DOMAIN) {
        Ok(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(true)])),
        _ => Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
    }
}
