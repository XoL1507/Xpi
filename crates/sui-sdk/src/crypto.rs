// Copyright (c) 2022, Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use rand::{rngs::StdRng, SeedableRng};
use serde::{Deserialize, Serialize};
use signature::Signer;
use std::collections::BTreeMap;
use std::fmt::Write;
use std::fmt::{Display, Formatter};
use std::fs;
use std::fs::File;
use std::io::BufReader;
use std::path::{Path, PathBuf};
use std::str::FromStr;

use bip39::Mnemonic;
use rand::rngs::adapter::ReadRng;

use sui_types::base_types::SuiAddress;
use sui_types::crypto::{
    get_key_pair_from_rng, EncodeDecodeBase64, PublicKey, Signature, SuiKeyPair,
};

#[derive(Serialize, Deserialize)]
#[non_exhaustive]
// This will work on user signatures, but not suitable for authority signatures.
pub enum KeystoreType {
    File(PathBuf),
    InMem(usize),
}

pub trait AccountKeystore: Send + Sync {
    fn sign(&self, address: &SuiAddress, msg: &[u8]) -> Result<Signature, signature::Error>;
    fn add_key(&mut self, keypair: SuiKeyPair) -> Result<(), anyhow::Error>;
    fn keys(&self) -> Vec<PublicKey>;
}

impl KeystoreType {
    pub fn init(&self) -> Result<SuiKeystore, anyhow::Error> {
        Ok(match self {
            KeystoreType::File(path) => SuiKeystore::from(FileBasedKeystore::load_or_create(path)?),
            KeystoreType::InMem(initial_key_number) => {
                SuiKeystore::from(InMemKeystore::new(*initial_key_number))
            }
        })
    }
}

impl Display for KeystoreType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut writer = String::new();
        match self {
            KeystoreType::File(path) => {
                writeln!(writer, "Keystore Type : File")?;
                write!(writer, "Keystore Path : {:?}", path)?;
                write!(f, "{}", writer)
            }
            KeystoreType::InMem(_) => {
                writeln!(writer, "Keystore Type : InMem")?;
                write!(f, "{}", writer)
            }
        }
    }
}

#[derive(Default)]
pub struct FileBasedKeystore {
    keys: BTreeMap<SuiAddress, SuiKeyPair>,
    path: Option<PathBuf>,
}

impl AccountKeystore for FileBasedKeystore {
    fn sign(&self, address: &SuiAddress, msg: &[u8]) -> Result<Signature, signature::Error> {
        self.keys
            .get(address)
            .ok_or_else(|| {
                signature::Error::from_source(format!("Cannot find key for address: [{address}]"))
            })?
            .try_sign(msg)
    }

    fn add_key(&mut self, keypair: SuiKeyPair) -> Result<(), anyhow::Error> {
        let address: SuiAddress = (&keypair.public()).into();
        self.keys.insert(address, keypair);
        self.save()?;
        Ok(())
    }

    fn keys(&self) -> Vec<PublicKey> {
        self.keys.values().map(|key| key.public()).collect()
    }
}

impl FileBasedKeystore {
    pub fn load_or_create(path: &Path) -> Result<Self, anyhow::Error> {
        let keys = if path.exists() {
            let reader = BufReader::new(File::open(path)?);
            let kp_strings: Vec<String> = serde_json::from_reader(reader)?;
            kp_strings
                .iter()
                .map(|kpstr| {
                    let key = SuiKeyPair::decode_base64(kpstr);
                    key.map(|k| (Into::<SuiAddress>::into(&k.public()), k))
                })
                .collect::<Result<BTreeMap<_, _>, _>>()
                .map_err(|_| anyhow::anyhow!("Invalid Keypair file"))?
        } else {
            BTreeMap::new()
        };

        Ok(Self {
            keys,
            path: Some(path.to_path_buf()),
        })
    }

    pub fn set_path(&mut self, path: &Path) {
        self.path = Some(path.to_path_buf());
    }

    pub fn save(&self) -> Result<(), anyhow::Error> {
        if let Some(path) = &self.path {
            let store = serde_json::to_string_pretty(
                &self
                    .keys
                    .values()
                    .map(EncodeDecodeBase64::encode_base64)
                    .collect::<Vec<_>>(),
            )
            .unwrap();
            fs::write(path, store)?
        }
        Ok(())
    }

    pub fn key_pairs(&self) -> Vec<&SuiKeyPair> {
        self.keys.values().collect()
    }
}

pub struct SuiKeystore(Box<dyn AccountKeystore>);

impl SuiKeystore {
    fn from<S: AccountKeystore + 'static>(keystore: S) -> Self {
        Self(Box::new(keystore))
    }

    pub fn add_key(&mut self, keypair: SuiKeyPair) -> Result<(), anyhow::Error> {
        self.0.add_key(keypair)
    }

    pub fn generate_new_key(&mut self) -> Result<(SuiAddress, String), anyhow::Error> {
        let mnemonic = Mnemonic::generate(12)?;
        let seed = mnemonic.to_seed("");
        let mut rng = RngWrapper(ReadRng::new(&seed));
        // TODO(joyqvq): add ability to generate as Secp256k1 keypair
        let (address, kp) = get_key_pair_from_rng(&mut rng);
        self.0.add_key(SuiKeyPair::Ed25519SuiKeyPair(kp))?;
        Ok((address, mnemonic.to_string()))
    }

    pub fn keys(&self) -> Vec<PublicKey> {
        self.0.keys()
    }

    pub fn addresses(&self) -> Vec<SuiAddress> {
        self.keys().iter().map(|k| k.into()).collect()
    }

    pub fn signer(&self, signer: SuiAddress) -> impl Signer<Signature> + '_ {
        KeystoreSigner::new(&*self.0, signer)
    }

    pub fn import_from_mnemonic(&mut self, phrase: &str) -> Result<SuiAddress, anyhow::Error> {
        let seed = &Mnemonic::from_str(phrase).unwrap().to_seed("");
        let mut rng = RngWrapper(ReadRng::new(seed));
        let (address, kp) = get_key_pair_from_rng(&mut rng);
        // TODO(joyqvq): add ability to import as Secp256k1 keypair
        self.0.add_key(SuiKeyPair::Ed25519SuiKeyPair(kp))?;
        Ok(address)
    }

    pub fn sign(&self, address: &SuiAddress, msg: &[u8]) -> Result<Signature, signature::Error> {
        self.0.sign(address, msg)
    }
}

/// wrapper for adding CryptoRng and RngCore impl to ReadRng.
struct RngWrapper<'a>(ReadRng<&'a [u8]>);

impl rand::CryptoRng for RngWrapper<'_> {}
impl rand::RngCore for RngWrapper<'_> {
    fn next_u32(&mut self) -> u32 {
        self.0.next_u32()
    }

    fn next_u64(&mut self) -> u64 {
        self.0.next_u64()
    }

    fn fill_bytes(&mut self, dest: &mut [u8]) {
        self.0.fill_bytes(dest)
    }

    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand::Error> {
        self.0.try_fill_bytes(dest)
    }
}

struct KeystoreSigner<'a> {
    keystore: &'a dyn AccountKeystore,
    address: SuiAddress,
}

impl<'a> KeystoreSigner<'a> {
    pub fn new(keystore: &'a dyn AccountKeystore, account: SuiAddress) -> Self {
        Self {
            keystore,
            address: account,
        }
    }
}

impl Signer<Signature> for KeystoreSigner<'_> {
    fn try_sign(&self, msg: &[u8]) -> Result<Signature, signature::Error> {
        self.keystore.sign(&self.address, msg)
    }
}

#[derive(Default)]
struct InMemKeystore {
    keys: BTreeMap<SuiAddress, SuiKeyPair>,
}

impl AccountKeystore for InMemKeystore {
    fn sign(&self, address: &SuiAddress, msg: &[u8]) -> Result<Signature, signature::Error> {
        self.keys
            .get(address)
            .ok_or_else(|| {
                signature::Error::from_source(format!("Cannot find key for address: [{address}]"))
            })?
            .try_sign(msg)
    }

    fn add_key(&mut self, keypair: SuiKeyPair) -> Result<(), anyhow::Error> {
        let address: SuiAddress = (&keypair.public()).into();
        self.keys.insert(address, keypair);
        Ok(())
    }

    fn keys(&self) -> Vec<PublicKey> {
        self.keys.values().map(|key| key.public()).collect()
    }
}

impl InMemKeystore {
    pub fn new(initial_key_number: usize) -> Self {
        let mut rng = StdRng::from_seed([0; 32]);
        let keys = (0..initial_key_number)
            .map(|_| get_key_pair_from_rng(&mut rng))
            .map(|(ad, k)| (ad, SuiKeyPair::Ed25519SuiKeyPair(k)))
            .collect::<BTreeMap<SuiAddress, SuiKeyPair>>();

        Self { keys }
    }
}

impl AccountKeystore for Box<dyn AccountKeystore> {
    fn sign(&self, address: &SuiAddress, msg: &[u8]) -> Result<Signature, signature::Error> {
        (**self).sign(address, msg)
    }

    fn add_key(&mut self, keypair: SuiKeyPair) -> Result<(), anyhow::Error> {
        (**self).add_key(keypair)
    }

    fn keys(&self) -> Vec<PublicKey> {
        (**self).keys()
    }
}
