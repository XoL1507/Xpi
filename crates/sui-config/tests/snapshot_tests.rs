// Copyright (c) 2022, Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

// This file contains tests that detect changes in Sui configs.
// If a PR breaks one or more tests here, the PR probably has a real impact
// on a production configuration file. When test failure happens, the PR should
// be marked as a breaking change and reviewers should be aware of this.
//
// Owners and operators of production configuration files can add themselves to
// .github/CODEOWNERS for the corresponding snapshot tests, so they can get notified
// of changes. PRs that modifies snapshot files should wait for reviews from
// code owners (if any) before merging.
//
// To review snapshot changes, and fix snapshot differences,
// 0. Install cargo-insta
// 1. Run `cargo insta test --review` under `./sui-config`.
// 2. Review, accept or reject changes.

use std::net::{IpAddr, Ipv4Addr, SocketAddr};
use std::path::PathBuf;

use insta::assert_yaml_snapshot;
use multiaddr::Multiaddr;
use narwhal_crypto::traits::KeyPair;
use rand::rngs::StdRng;
use rand::SeedableRng;
use sui_config::{genesis::Builder, genesis_config::GenesisConfig};
use sui_config::{NetworkConfig, ValidatorInfo};
use sui_types::base_types::{ObjectID, SuiAddress};
use sui_types::crypto::{get_key_pair_from_rng, AuthorityKeyPair, AuthorityPublicKeyBytes};

#[test]
fn genesis_config_snapshot_matches() {
    // Test creating fake SuiAddress from PublicKeyBytes.
    let keypair: AuthorityKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let public_key = AuthorityPublicKeyBytes::from(keypair.public());
    let fake_addr = SuiAddress::from(&public_key);

    let fake_obj_id = ObjectID::from(fake_addr);
    let mut genesis_config = GenesisConfig::for_local_testing();
    for account in &mut genesis_config.accounts {
        account.address = Some(fake_addr);
        for gas_obj in &mut account.gas_objects {
            gas_obj.object_id = fake_obj_id;
        }
    }
    assert_yaml_snapshot!(genesis_config);
}

// If a Move change breaks the test below, it is because related Move functions are included
// in the genesis blob. The test failure can be fixed by updating the snapshot / blob.
// Also, folks in the file's CODEWOWNERS should be notified about the change. They may need to
// re-create the genesis blob.
#[test]
fn empty_genesis_snapshot_matches() {
    let genesis = Builder::new().build();
    assert_yaml_snapshot!(genesis);
}

#[test]
fn populated_genesis_snapshot_matches() {
    let genesis_config = GenesisConfig::for_local_testing();
    let (_account_keys, objects) = genesis_config
        .generate_accounts(&mut StdRng::from_seed([0; 32]))
        .unwrap();
    let key: AuthorityKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let validator = ValidatorInfo {
        name: "0".into(),
        public_key: key.public().into(),
        stake: 1,
        delegation: 0,
        gas_price: 1,
        network_address: Multiaddr::empty(),
        narwhal_primary_to_primary: Multiaddr::empty(),
        narwhal_worker_to_primary: Multiaddr::empty(),
        narwhal_primary_to_worker: Multiaddr::empty(),
        narwhal_worker_to_worker: Multiaddr::empty(),
        narwhal_consensus_address: Multiaddr::empty(),
    };

    let genesis = Builder::new()
        .add_objects(objects)
        .add_validator(validator)
        .build();
    assert_yaml_snapshot!(genesis.validator_set());
    assert_yaml_snapshot!(genesis.committee().unwrap());
    assert_yaml_snapshot!(genesis.narwhal_committee());
    assert_yaml_snapshot!(genesis.sui_system_object());
    // Serialized `genesis` is not static and cannot be snapshot tested.
}

#[test]
fn network_config_snapshot_matches() {
    let temp_dir = tempfile::tempdir().unwrap();
    let committee_size = 7;
    let rng = StdRng::from_seed([0; 32]);
    let mut network_config = NetworkConfig::generate_with_rng(temp_dir.path(), committee_size, rng);
    // TODO: Inject static temp path and port numbers, instead of clearing them.
    for mut validator_config in &mut network_config.validator_configs {
        validator_config.db_path = PathBuf::from("/tmp/foo/");
        validator_config.network_address = Multiaddr::empty();
        let fake_socket = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 1);
        validator_config.json_rpc_address = fake_socket;
        validator_config.metrics_address = fake_socket;
        validator_config.admin_interface_port = 8888;
        let metrics_addr: Multiaddr = "/ip4/127.0.0.1/tcp/1234".parse().unwrap();
        if let Some(consensus_config) = validator_config.consensus_config.as_mut() {
            consensus_config.consensus_address = Multiaddr::empty();
            consensus_config.consensus_db_path = PathBuf::from("/tmp/foo/");
            consensus_config
                .narwhal_config
                .consensus_api_grpc
                .socket_addr = Multiaddr::empty();
            consensus_config
                .narwhal_config
                .prometheus_metrics
                .socket_addr = metrics_addr
        }
    }
    assert_yaml_snapshot!(network_config, {
        ".genesis" => "[fake genesis]",
        ".validator_configs[].genesis.genesis" => "[fake genesis]",
    });
}
