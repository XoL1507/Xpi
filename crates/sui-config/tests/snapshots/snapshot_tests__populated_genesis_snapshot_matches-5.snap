---
source: crates/sui-config/tests/snapshot_tests.rs
expression: genesis.sui_system_object()
---
info:
  id: "0x0000000000000000000000000000000000000005"
epoch: 0
validators:
  validator_stake: 1
  delegation_stake: 0
  quorum_stake_threshold: 1
  active_validators:
    - metadata:
        sui_address: 375741c9373ae27bde5e6b5f5cc79f074cd84587
        pubkey_bytes:
          - 140
          - 102
          - 220
          - 44
          - 30
          - 169
          - 229
          - 63
          - 9
          - 133
          - 193
          - 123
          - 78
          - 122
          - 241
          - 153
          - 18
          - 182
          - 211
          - 196
          - 14
          - 12
          - 89
          - 32
          - 165
          - 161
          - 37
          - 9
          - 180
          - 235
          - 54
          - 25
          - 245
          - 224
          - 126
          - 197
          - 110
          - 167
          - 127
          - 11
          - 48
          - 98
          - 155
          - 161
          - 204
          - 114
          - 215
          - 91
          - 19
          - 148
          - 96
          - 120
          - 42
          - 95
          - 14
          - 47
          - 137
          - 251
          - 76
          - 66
          - 180
          - 184
          - 165
          - 250
          - 227
          - 210
          - 96
          - 16
          - 34
          - 32
          - 230
          - 61
          - 7
          - 84
          - 231
          - 225
          - 132
          - 109
          - 238
          - 253
          - 57
          - 136
          - 234
          - 222
          - 78
          - 211
          - 127
          - 19
          - 133
          - 67
          - 125
          - 25
          - 222
          - 26
          - 22
          - 24
        network_pubkey_bytes:
          - 117
          - 99
          - 98
          - 117
          - 70
          - 106
          - 68
          - 118
          - 80
          - 110
          - 69
          - 82
          - 82
          - 75
          - 90
          - 73
          - 50
          - 119
          - 97
          - 55
          - 115
          - 105
          - 104
          - 80
          - 99
          - 110
          - 84
          - 80
          - 118
          - 117
          - 85
          - 47
          - 47
          - 79
          - 53
          - 81
          - 80
          - 77
          - 71
          - 107
          - 107
          - 103
          - 65
          - 61
        proof_of_possession_bytes:
          - 139
          - 91
          - 16
          - 21
          - 252
          - 177
          - 83
          - 235
          - 179
          - 45
          - 51
          - 27
          - 32
          - 108
          - 191
          - 122
          - 34
          - 233
          - 8
          - 191
          - 174
          - 39
          - 53
          - 225
          - 71
          - 185
          - 190
          - 152
          - 161
          - 188
          - 205
          - 164
          - 114
          - 241
          - 218
          - 40
          - 154
          - 30
          - 255
          - 169
          - 124
          - 94
          - 223
          - 87
          - 68
          - 128
          - 197
          - 132
        name:
          - 48
        net_address: []
        next_epoch_stake: 1
        next_epoch_delegation: 0
        next_epoch_gas_price: 1
      stake_amount: 1
      pending_stake: 0
      pending_withdraw: 0
      gas_price: 1
      delegation_staking_pool:
        validator_address: 375741c9373ae27bde5e6b5f5cc79f074cd84587
        starting_epoch: 1
        epoch_starting_sui_balance: 0
        epoch_starting_delegation_token_supply: 0
        sui_balance: 0
        rewards_pool:
          value: 0
        delegation_token_supply:
          value: 0
        pending_delegations: []
  pending_validators: []
  pending_removals: []
  next_epoch_validators:
    - sui_address: 375741c9373ae27bde5e6b5f5cc79f074cd84587
      pubkey_bytes:
        - 140
        - 102
        - 220
        - 44
        - 30
        - 169
        - 229
        - 63
        - 9
        - 133
        - 193
        - 123
        - 78
        - 122
        - 241
        - 153
        - 18
        - 182
        - 211
        - 196
        - 14
        - 12
        - 89
        - 32
        - 165
        - 161
        - 37
        - 9
        - 180
        - 235
        - 54
        - 25
        - 245
        - 224
        - 126
        - 197
        - 110
        - 167
        - 127
        - 11
        - 48
        - 98
        - 155
        - 161
        - 204
        - 114
        - 215
        - 91
        - 19
        - 148
        - 96
        - 120
        - 42
        - 95
        - 14
        - 47
        - 137
        - 251
        - 76
        - 66
        - 180
        - 184
        - 165
        - 250
        - 227
        - 210
        - 96
        - 16
        - 34
        - 32
        - 230
        - 61
        - 7
        - 84
        - 231
        - 225
        - 132
        - 109
        - 238
        - 253
        - 57
        - 136
        - 234
        - 222
        - 78
        - 211
        - 127
        - 19
        - 133
        - 67
        - 125
        - 25
        - 222
        - 26
        - 22
        - 24
      network_pubkey_bytes:
        - 117
        - 99
        - 98
        - 117
        - 70
        - 106
        - 68
        - 118
        - 80
        - 110
        - 69
        - 82
        - 82
        - 75
        - 90
        - 73
        - 50
        - 119
        - 97
        - 55
        - 115
        - 105
        - 104
        - 80
        - 99
        - 110
        - 84
        - 80
        - 118
        - 117
        - 85
        - 47
        - 47
        - 79
        - 53
        - 81
        - 80
        - 77
        - 71
        - 107
        - 107
        - 103
        - 65
        - 61
      proof_of_possession_bytes:
        - 139
        - 91
        - 16
        - 21
        - 252
        - 177
        - 83
        - 235
        - 179
        - 45
        - 51
        - 27
        - 32
        - 108
        - 191
        - 122
        - 34
        - 233
        - 8
        - 191
        - 174
        - 39
        - 53
        - 225
        - 71
        - 185
        - 190
        - 152
        - 161
        - 188
        - 205
        - 164
        - 114
        - 241
        - 218
        - 40
        - 154
        - 30
        - 255
        - 169
        - 124
        - 94
        - 223
        - 87
        - 68
        - 128
        - 197
        - 132
      name:
        - 48
      net_address: []
      next_epoch_stake: 1
      next_epoch_delegation: 0
      next_epoch_gas_price: 1
treasury_cap:
  value: 100000000000001
storage_fund:
  value: 100000000000000
parameters:
  min_validator_stake: 100000000000000
  max_validator_candidate_count: 100
  storage_gas_price: 1
reference_gas_price: 0
validator_report_records:
  contents: []

